src/Microsoft.Data/PrimitiveColumnComputations.cs

```diff
@@ -0,0 +1,1864 @@
+﻿
+
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+// Generated from PrimitiveColumnComputations.tt. Do not modify directly
+
+using System;
+
+namespace Microsoft.Data
+{
+    internal interface IPrimitiveColumnComputation<T>
+        where T : struct
+    {
+        void Abs(PrimitiveColumnContainer<T> column);
+        void All(PrimitiveColumnContainer<T> column, out bool ret);
+        void Any(PrimitiveColumnContainer<T> column, out bool ret);
+        void Cummax(PrimitiveColumnContainer<T> column);
+        void Cummin(PrimitiveColumnContainer<T> column);
+        void Cumprod(PrimitiveColumnContainer<T> column);
+        void Cumsum(PrimitiveColumnContainer<T> column);
+        void Max(PrimitiveColumnContainer<T> column, out T ret);
+        void Min(PrimitiveColumnContainer<T> column, out T ret);
+        void Prod(PrimitiveColumnContainer<T> column, out T ret);
+        void Sum(PrimitiveColumnContainer<T> column, out T ret);
+        void Round(PrimitiveColumnContainer<T> column);
+    }
+
+    internal static class PrimitiveColumnComputation<T>
+        where T : struct
+    {
+        public static IPrimitiveColumnComputation<T> Instance => PrimitiveColumnComputation.GetComputation<T>();
```

Comment:
This will new up the Computation object every time it is called. That seems wasteful. Instead, we should cache it in a field.

Reply:
Really? Because this is static, shouldn't this new up a computation object once per `T` and reuse that for next time?

Reply:
>  shouldn't this new up a computation object once per T and reuse that for next time?

I would think that is the behavior we want, but the code you've written doesn't do that.

```C#
public static IPrimitiveColumnComputation<T> Instance => PrimitiveColumnComputation.GetComputation<T>();
```

That is short-hand for:

```C#
public static IPrimitiveColumnComputation<T> Instance
{
    get { return PrimitiveColumnComputation.GetComputation<T>(); }
}
```

Which means every time the property is invoked/called it will call `GetComputation<T>()`, which in turn returns a new object every time it is invoked.

Reply:
`=>` means execute the following code each time. `=` means assignment. You need the last one to initialize the member only once when the static constructor is called.

```suggestion
        public static IPrimitiveColumnComputation<T> Instance = PrimitiveColumnComputation.GetComputation<T>();
```

Reply:
However, the suggested change is turning `Instance` from a property into a field. And as written above, allows the field to be publicly settable.

In general, we shouldn't be exposing fields publicly unless there is a strong reason for it. We should keep it a read-only property. But it should reuse the same object each time it is invoked, so it doesn't create unnecessary allocations.

Reply:
Yep, fixed. Thanks!

Should this rule be applied for internal classes?

Reply:
I missed that it was an internal class. That does change the "property or field" decision.

It is probably personal preference, but I generally tend to keep all fields (static or instance) private by default and exposed through properties. And only expose them as fields if there are specific reasons for it.